query-planner-js/src/Scope.ts

```diff
@@ -0,0 +1,271 @@
+    if (!this.cachedRuntimeTypes) {
+      this.cachedRuntimeTypes = this.computePossibleRuntimeTypes();
+    }
+    return this.cachedRuntimeTypes;
+  }
+
+  private static valueIdentityKey(value: ValueNode) : string {
+    // We distinguish different value types though a leading character with a comma (so the int 3 has key "i'3") to
+    // avoid conflicts.
+    switch (value.kind) {
+      case 'Variable':
+        return value.name.value;
+      case 'IntValue':
+        return "i'" + value.value;
+      case 'FloatValue':
+        return "f'" + value.value;
+      case 'EnumValue':
+        return "e'" + value.value;
+      case 'StringValue':
+        return "s'" + value.value;
+      case 'BooleanValue':
+        return "b'" + String(value.value);
+      case 'NullValue':
+        return "<null>";
+      case 'ListValue':
+        return "[" + value.values.map(this.valueIdentityKey).join('-') + "]";
+      case 'ObjectValue':
+        const fields = value.fields.map(f => f.name.value + '-' + this.valueIdentityKey(f.value));
+        fields.sort(); // Field order is not semantically significant in graphQL.
+        return "{" + fields.join('-') + "}";
+    }
+  }
+
+  private static directiveIdentityKey(directive: DirectiveNode) : string {
+    const argsKeys = directive.arguments
+      ? directive.arguments.map(arg => arg.name.value + '-' + Scope.valueIdentityKey(arg.value))
+      : [];
+    argsKeys.sort(); // Argument order is not semantically significant in graphQL.
+    return `${directive.name.value}-${argsKeys.join('-')}`;
+  }
+
+  private static directivesIdentityKey(directives: ReadonlyArray<DirectiveNode>) : string {
+    const keys = directives.map(d => Scope.directiveIdentityKey(d));
+    // We sort the directives keys as we don't want the order application of directives to matter
+    // in the equality. In other words, we want:
+    //   ... on X @foo @bar
+    // to be the same as:
+    //   ... on X @bar @foo
+    keys.sort();
+    return keys.join('-');
+  }
+
+  private computeIdentityKey(): string {
+      const directivesKey = this.directives ? Scope.directivesIdentityKey(this.directives) : "";
+      const enclosingKey = this.enclosing ? this.enclosing.computeIdentityKey() : "";
+      return `${this.parentType}-${directivesKey}-${enclosingKey}`;
+  }
+
+  /**
+   * A string value that uniquely identify the scope.
+   *
+   * The "identity key" of 2 scope objects can be tested for equality to decide if 2 scopes are equal. This exists
+   * so that scopes can be (kind of) used as map keys: javacript maps always only use reference equality for objects
+   * when used as key, so using this string allows to effectively get value equality.
+   *
+   * Note that the main property this method must ensure is that if 2 scopes have the same `identityKey()`, then they
+   * are semantically equilvalent (their treatment by the query planner is undistiguishable). If 2 scopes are
```

Comment:
I don't follow what the second sentence is saying here.

The first sentence says "same IK implies equivalent".

The second sentence says "equivalent implies different IK".

Is there a typo in one of the sentences?

Reply:
There was a typo. Fixed.